File: lib/models/medication.dart
class Medication {
  final String id;
  final String userId;
  final String drugName;
  final String dosage;
  final String frequency;
  final String whenToTake;
  final String prescribedBy;
  final DateTime startDate;
  final DateTime? endDate;
  final String status; // 'active' | 'stopped'
  final String stopReason;
  final String notes;
  final List<DateTime> takenDates; // ISO timestamps of doses taken

  Medication({
    required this.id,
    required this.userId,
    required this.drugName,
    required this.dosage,
    required this.frequency,
    this.whenToTake = 'With meals',
    this.prescribedBy = '',
    required this.startDate,
    this.endDate,
    this.status = 'active',
    this.stopReason = '',
    this.notes = '',
    this.takenDates = const [],
  });

  bool get isActive => status.toLowerCase() == 'active';

  bool takenToday() {
    final now = DateTime.now();
    return takenDates.any((d) =>
      d.year == now.year && d.month == now.month && d.day == now.day
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'userId': userId,
      'drugName': drugName,
      'dosage': dosage,
      'frequency': frequency,
      'whenToTake': whenToTake,
      'prescribedBy': prescribedBy,
      'startDate': startDate.toIso8601String(),
      'endDate': endDate?.toIso8601String(),
      'status': status,
      'stopReason': stopReason,
      'notes': notes,
      'takenDates': takenDates.map((d) => d.toIso8601String()).toList(),
    };
  }

  factory Medication.fromMap(Map<dynamic, dynamic> map) {
    return Medication(
      id: map['id'] ?? '',
      userId: map['userId'] ?? '',
      drugName: map['drugName'] ?? '',
      dosage: map['dosage'] ?? '',
      frequency: map['frequency'] ?? '',
      whenToTake: map['whenToTake'] ?? 'With meals',
      prescribedBy: map['prescribedBy'] ?? '',
      startDate: map['startDate'] != null ? DateTime.parse(map['startDate']) : DateTime.now(),
      endDate: map['endDate'] != null ? DateTime.parse(map['endDate']) : null,
      status: map['status'] ?? 'active',
      stopReason: map['stopReason'] ?? '',
      notes: map['notes'] ?? '',
      takenDates: (map['takenDates'] as List? ?? [])
          .map((d) => DateTime.parse(d.toString()))
          .toList(),
    );
  }
}

File: backend/routes/medication.py
"""Medication CRUD routes."""
from datetime import datetime
from flask import Blueprint, request, jsonify, g
from extensions import db
from models import Medication
from middleware import token_required

medication_bp = Blueprint('medication', __name__, url_prefix='/api/prescription')


@medication_bp.route('/medications', methods=['GET'])
@token_required
def list_medications():
    user = g.current_user
    status = request.args.get('status')

    query = Medication.query.filter_by(user_id=user.id)
    if status:
        query = query.filter_by(status=status)

    meds = query.order_by(Medication.created_at.desc()).all()
    return jsonify([m.to_dict() for m in meds]), 200


@medication_bp.route('/medications', methods=['POST'])
@token_required
def add_medication():
    user = g.current_user
    data = request.get_json()
    if not data:
        return jsonify({'error': 'Request body is required.'}), 400

    drug_name = (data.get('drug_name') or '').strip()
    if not drug_name:
        return jsonify({'error': 'Drug name is required.'}), 400

    med = Medication(
        user_id=user.id,
        drug_name=drug_name,
        dosage=(data.get('dosage') or '').strip(),
        frequency=(data.get('frequency') or '').strip(),
        when_to_take=(data.get('when_to_take') or '').strip(),
        prescribed_by=(data.get('prescribed_by') or '').strip(),
        start_date=datetime.utcnow(),
        status='active',
        notes=(data.get('notes') or '').strip(),
        duration=(data.get('duration') or '').strip(),
    )
    db.session.add(med)
    db.session.commit()

    return jsonify(med.to_dict()), 201


@medication_bp.route('/medications/<int:med_id>', methods=['PUT'])
@token_required
def update_medication(med_id):
    user = g.current_user
    med = Medication.query.filter_by(id=med_id, user_id=user.id).first()
    if not med:
        return jsonify({'error': 'Medication not found.'}), 404

    data = request.get_json()
    if not data:
        return jsonify({'error': 'Request body is required.'}), 400

    if 'drug_name' in data:
        med.drug_name = data['drug_name']
    if 'dosage' in data:
        med.dosage = data['dosage']
    if 'frequency' in data:
        med.frequency = data['frequency']
    if 'when_to_take' in data:
        med.when_to_take = data['when_to_take']
    if 'prescribed_by' in data:
        med.prescribed_by = data['prescribed_by']
    if 'status' in data:
        med.status = data['status']
        if data['status'] == 'stopped':
            med.end_date = datetime.utcnow()
    if 'stop_reason' in data:
        med.stop_reason = data['stop_reason']
    if 'notes' in data:
        med.notes = data['notes']

    db.session.commit()
    return jsonify(med.to_dict()), 200


@medication_bp.route('/medications/<int:med_id>', methods=['DELETE'])
@token_required
def delete_medication(med_id):
    user = g.current_user
    med = Medication.query.filter_by(id=med_id, user_id=user.id).first()
    if not med:
        return jsonify({'error': 'Medication not found.'}), 404

    db.session.delete(med)
    db.session.commit()
    return jsonify({'message': 'Medication deleted.'}), 200
